File: projects/python-sdk/handlers/v4/fingerprint_client.py
import json
from typing import Callable

from pydantic import ValidationError
from fingerprint_server_sdk import FingerprintApi, Configuration
from fingerprint_server_sdk.api_response import ApiResponse
from fingerprint_server_sdk.configuration import Region
from fingerprint_server_sdk.rest import ApiException
from flask import request, jsonify


def create_client():
    api_key = request.args.get('api_key', '')
    region = request.args.get('region')
    # Bypass client validation to get Server API v4 error response
    try:
        resolved_region = Region(region) if region else None
    except ValueError:
        resolved_region = None
    configuration = Configuration(api_key=api_key, region=resolved_region)
    return FingerprintApi(configuration)


def prepare_response(data, code, raw_data):
    parsed = data.to_dict() if data is not None else data
    return {
        "code": code,
        "originalResponse": str(raw_data),
        "parsedResponse": parsed,
    }


def prepare_error_response(error: Exception):
    if isinstance(error, ApiException):
        try:
            parsed = json.loads(error.body) if error.body else str(error.reason)
        except (json.JSONDecodeError, TypeError):
            parsed = str(error.reason)
        return {
            "code": error.status,
            "originalResponse": str(error),
            "parsedResponse": parsed,
        }

    # Bypass client validation and mock Server API v4 error response
    if isinstance(error, ValidationError):
        first_error = error.errors()[0]
        field = first_error['loc'][-1]
        field_messages = {
            "bot": "invalid bot type",
        }
        message = field_messages.get(field, f"invalid {field}")
        return {
            "code": 400,
            "originalResponse": str(error),
            "parsedResponse": {
                "error": {
                    "code": "request_cannot_be_parsed",
                    "message": message,
                }
            },
        }

    return {
        "code": 500,
        "originalResponse": str(error),
        "parsedResponse": str(error),
    }


def create_response(api_call: Callable[[], ApiResponse]):
    try:
        api_response = api_call()
        return jsonify(prepare_response(api_response.data, api_response.status_code, api_response.raw_data))
    except Exception as e:
        return jsonify(prepare_error_response(e))

File: projects/php-sdk/src/V4/FingerprintClient.php
<?php

namespace PHP_SDK\V4;

use Fingerprint\ServerSdk\Api\FingerprintApi;
use Fingerprint\ServerSdk\ApiException;
use Fingerprint\ServerSdk\Configuration;
use GuzzleHttp\Psr7\Response;
use InvalidArgumentException;
use Psr\Http\Message\MessageInterface;
use Throwable;
use ValueError;

class FingerprintClient
{
    private const FIELD_MESSAGES = [
        'bot' => 'invalid bot type',
    ];

    public static function createFromQueryParams(array $queryParams): FingerprintApi
    {
        $apiKey = $queryParams['api_key'] ?? '';
        $region = $queryParams['region'] ?? null;

        return FingerprintClient::create($apiKey, $region);
    }

    public static function create(string $apiKey, ?string $region = null): FingerprintApi
    {
        $config = new Configuration($apiKey);
        if ($region) {
            $config->setRegion($region);
        }

        return new FingerprintApi($config);
    }

    public static function createResponse(Response $response, callable $apiCall): MessageInterface
    {
        try {
            list($model, $httpResponse) = $apiCall();
            $result = [
                'code' => $httpResponse->getStatusCode(),
                'originalResponse' => (string) $httpResponse->getBody(),
                'parsedResponse' => $model,
            ];
        } catch (ApiException $e) {
            $result = [
                'code' => $e->getCode(),
                'originalResponse' => (string) $e->getResponseObject()?->getBody(),
                'parsedResponse' => $e->getErrorDetails(),
            ];
        } catch (InvalidArgumentException|ValueError $e) {
            $result = self::buildValidationErrorResult($e);
        } catch (Throwable $e) {
            $result = [
                'code' => 500,
                'originalResponse' => $e->getMessage(),
                'parsedResponse' => $e->getMessage(),
            ];
        }

        $response->getBody()->write(json_encode($result));

        return $response->withHeader('Content-Type', 'application/json');
    }

    public static function buildValidationErrorResult(Throwable $e): array
    {
        $message = $e->getMessage();
        $field = null;

        // Mock parameter validation
        if (preg_match('/invalid value for "\$(\w+)"/', $message, $matches)) {
            $field = $matches[1];
        }

        // Mock enum validation (for searchEvents only)
        elseif (preg_match('/"[^"]+" is not a valid backing value for enum .*\\\\(\w+)$/', $message, $matches)) {
            $enumClass = $matches[1];
            $field = preg_replace('/^SearchEvents/', '', $enumClass);
            $field = strtolower(preg_replace('/([a-z])([A-Z])/', '$1_$2', $field));
        }

        if ($field) {
            $message = self::FIELD_MESSAGES[$field] ?? 'invalid ' . str_replace('_', ' ', $field);
        }

        return [
            'code' => 400,
            'originalResponse' => $e->getMessage(),
            'parsedResponse' => [
                'error' => [
                    'code' => 'request_cannot_be_parsed',
                    'message' => $message,
                ],
            ],
        ];
    }
}
